Create development environment

caseycesari opened the pull request:
## Overview

Sets up a development environment based on create-react-app, adds support for SASS, and other components of our typical project setup. Also demos that Redux is wired up correctly by implementing the counter demo from the app-template.

Connects #6 

### Demo

![dec-11-2018 15-08-58](https://user-images.githubusercontent.com/1042475/49827193-bf1a8080-fd56-11e8-86e1-e85482aaee63.gif)

### Notes

Follow-up issues: #7, #8 

## Testing Instructions

- Run `./scripts/setup`.
- Run `./scripts/server`.
- Visit the project in the browser.
- Verify that clicking the button increases the counter via Redux.

caseycesari commented on README.md:
```diff
@@ -1,2 +1,59 @@
+
+### Requirements
+
+* Yarn
```
I thought about adding a Docker image, but wasn't sure if people wanted to try working with CRA in it's rawest form or from within Docker.

@mmcfarland what do you think?

caseycesari replied:
Though I suppose a Vagrant VM would be more appropriate here.

caseycesari replied:
Sounds good. I'll add Vagrant + Docker.

Re: `yarn`, I went with that because it appears to be the default of CRA, but `npm` is also supported. I can make sure `npm` does indeed work.

kellyi replied:
If it's in Vagrant + Docker, no need to make npm work. The Docker Node image has yarn installed by default.

caseycesari commented on package.json:
```diff
@@ -0,0 +1,33 @@
+    "redux": "^4.0.1",
+    "redux-act": "^1.7.4",
+    "redux-logger": "^3.0.6",
+    "redux-thunk": "^2.3.0"
```
Yes, both are included. We use `prop-types` [here](https://github.com/azavea/ism-watershed-wellness-snapshot/pull/9/commits/4b4256000743d35783211e833bb77c751b459d94#diff-14b1e33d5bf5649597cdc0e4f684daddR2). And here's some [documentation](https://github.com/facebook/create-react-app/tree/master/packages/eslint-config-react-app) on the ESLint config that comes along with CRA.

kellyi commented on src/store.js:
```diff
@@ -0,0 +1,15 @@
+const middlewares = [thunk];
+
+if (process.env.NODE_ENV === 'development') {
+    const logger = createLogger();
```
To improve the output we should initialize the logger using `createLogger({ diff: true, collapsed: true })`; this will collapse the actions by default and also output a diff of state changes.

caseycesari replied:
Thanks. Will do.

kellyi commented on src/index.js:
```diff
@@ -0,0 +1,31 @@
+import reducers from './reducers';
+
+const store = createStoreWithMiddleware(reducers);
+assignAll(actions, store)
```
I checked it doesn't seem like actions bound with `assignAll` can be written using the thunk syntax we typically use for API requests. I don't think we need to drop or change this now but keeping it will eventually require segmenting actions into files based on

1. whether or not they are thunks
2. whether they accept multiple arguments or need to be curried or before dispatch
3. whether they might need internal access to `getState` to check on some value in the store

...which adds back some of the complexity we're trying to circumvent with the `assignAll` syntax.

I think I generally prefer using `mapDispatchToProps` and `bindActionCreators` -- https://redux.js.org/api/bindactioncreators -- to `assignAll` syntax, but it may be that this application's level of complexity works with `assignAll`.

Anyway: fine to leave as is, just thinking through the implications.

caseycesari replied:
Thanks for the overview. I prefer `assignAll` for its simplicity, but realize it might not be able to handle everything we need. It worked for Urban Water Revealed, but that was a fairly simple app. I'll keep this for now, but open to changes when we need them.

kellyi commented on src/App.js:
```diff
@@ -0,0 +1,51 @@
+            Edit <code>src/App.js</code> and save to reload.
+          </p>
+          <a
+            className="App-link"
```
Sounds good.

kellyi commented:
I set this up locally, btw, and it worked. Seems like the remaining work here is to place it in a Vagrant VM + Docker.

caseycesari commented:
Added a VM and Docker container, and updated the scripts accordingly. Ready for another look.

caseycesari commented:
Re: `/scripts/lint`, I'm going to punt on that until integrating prettier, since that will have to be considered, and it doesn't immediately appear to be that simple to run eslint from the command line with CRA.

kellyi commented:
Setting this up now!

kellyi commented on docker-compose.yml:
```diff
@@ -0,0 +1,13 @@
+version: '3'
+services:
+  app:
+    image: node:8.11
```
I think we should move up to Node 10 which is the current LTS. Consider also not having the pin to the minor version here, just `node:10`. We could also probably use alpine or slim here.

caseycesari replied:
Makes sense. I'll switch to `node:10`

caseycesari replied:
Updated.

kellyi commented:
Got this set up and the initial set up is working. However, I do not have hot reloading or anything on file system changes -- I think because the fs changes may not thread through Vagrant and Docker?

caseycesari commented:
Ready for another look. Fixed the filesystem by making the changes suggested here: https://github.com/azavea/open-apparel-registry/pull/11#discussion_r240812162.

As we discussed, hot-reloading is not on by default. But I added an issue to figure it out.
